// rclone/dist/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultSalt = exports.blockSize = exports.blockDataSize = exports.blockHeaderSize = exports.fileHeaderSize = exports.fileNonceSize = exports.fileMagicSize = exports.fileMagic = exports.keySize = void 0;
// Constants
var keySize = 32 + 32 + 16; // File Constants

exports.keySize = keySize;
var fileMagic = 'RCLONE\x00\x00';
exports.fileMagic = fileMagic;
var fileMagicSize = fileMagic.length;
exports.fileMagicSize = fileMagicSize;
var fileNonceSize = 24;
exports.fileNonceSize = fileNonceSize;
var fileHeaderSize = fileMagicSize + fileNonceSize;
exports.fileHeaderSize = fileHeaderSize;
var blockHeaderSize = 16; // crypto_secretbox_BOXZEROBYTES

exports.blockHeaderSize = blockHeaderSize;
var blockDataSize = 64 * 1024;
exports.blockDataSize = blockDataSize;
var blockSize = blockHeaderSize + blockDataSize;
exports.blockSize = blockSize;
var defaultSalt = [0xa8, 0x0d, 0xf4, 0x3a, 0x8f, 0xbd, 0x03, 0x08, 0xa7, 0xca, 0xb8, 0x3e, 0x58, 0x1f, 0x86, 0xb1];
exports.defaultSalt = defaultSalt;
